packages/modules/order/src/services/order-module-service.ts

```diff
@@ -2302,6 +2303,132 @@ export default class OrderModuleService<
+      },
+      sharedContext
+    )
+
+    if (order.version < 2) {
+      throw new MedusaError(
+        MedusaError.Types.INVALID_DATA,
+        `Order with id ${orderId} has no previous versions`
+      )
+    }
+
+    return await this.undoLastChange_(order, lastOrderChange, sharedContext)
+  }
+
+  @InjectTransactionManager()
+  protected async undoLastChange_(
+    order: OrderDTO,
+    lastOrderChange?: Partial<OrderChangeDTO>,
+    sharedContext?: Context
+  ): Promise<void> {
+    const currentVersion = order.version
+
+    // Order Changes
+    const orderChanges = await this.orderChangeService_.list(
+      {
+        order_id: order.id,
+        version: currentVersion,
+      },
+      { select: ["id", "version"] },
+      sharedContext
+    )
+
+    const orderChangesIds = orderChanges.map((change) => {
+      return {
+        id: change.id,
+        status: lastOrderChange?.status ?? OrderChangeStatus.PENDING,
+        confirmed_at: null,
+      }
+    })
+
+    await this.orderChangeService_.update(orderChangesIds, sharedContext)
+
+    // Order Changes Actions
+    const orderChangesActions = await this.orderChangeActionService_.list(
+      {
+        order_id: order.id,
+        version: currentVersion,
+      },
+      { select: ["id", "version"] },
+      sharedContext
+    )
+    const orderChangeActionsIds = orderChangesActions.map((action) => {
+      return {
+        id: action.id,
+        applied: false,
+      }
+    })
+
+    await this.orderChangeActionService_.update(
+      orderChangeActionsIds,
+      sharedContext
+    )
+
+    // Order Summary
+    const orderSummary = await this.orderSummaryService_.list(
+      {
+        order_id: order.id,
+        version: currentVersion,
+      },
+      { select: ["id", "version"] },
+      sharedContext
+    )
+    const orderSummaryIds = orderSummary.map((summary) => summary.id)
+
+    await this.orderSummaryService_.softDelete(orderSummaryIds, sharedContext)
+
+    // Order Items
+    const orderItems = await this.orderItemService_.list(
+      {
+        order_id: order.id,
+        version: currentVersion,
+      },
+      { select: ["id", "version"] },
+      sharedContext
+    )
+    const orderItemIds = orderItems.map((summary) => summary.id)
+
+    await this.orderItemService_.softDelete(orderItemIds, sharedContext)
+
+    // Order Shipping
+    const orderShippings = await this.orderShippingService_.list(
+      {
+        order_id: order.id,
+        version: currentVersion,
+      },
+      { select: ["id", "version"] },
+      sharedContext
+    )
+    const orderShippingIds = orderShippings.map((sh) => sh.id)
+
+    await this.orderShippingService_.softDelete(orderShippingIds, sharedContext)
+
+    // Order
+    await this.orderService_.update(
+      {
+        selector: {
+          id: order.id,
+        },
+        data: {
+          version: order.version - 1,
+        },
+      },
+      sharedContext
+    )
+  }
```

Comment:
Q: is there any of the update that could be concurrently done or do they rely on each other?

Reply:
All of the can be in parallel or most of them. Same with revertLastChange method.
I will update it, even though mikro-orm is doing that under the hood anyway.

Reply:
yeah I was considering that mikro orm does it fo us, but just from an implementation pov I prefer to be explicit in case for some reason a flush happen in the middle or tomorrow people base their impl on that while being on an external module using another tool that does not do it for them, that way they can see best practices :)